src/background-view.cpp

```diff
@@ -0,0 +1,355 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <map>
+#include <fcntl.h>
+#include <unistd.h>
+#include <signal.h>
+#include <sys/wait.h>
+#include <sys/socket.h>
+#include "wayfire/core.hpp"
+#include "wayfire/view.hpp"
+#include "wayfire/plugin.hpp"
+#include "wayfire/output.hpp"
+#include "wayfire/signal-definitions.hpp"
+#include "wayfire/workspace-manager.hpp"
+#include "wayfire/output-layout.hpp"
+
+#include <wayfire/util/log.hpp>
+
+extern "C"
+{
+#include <wlr/config.h>
+#if WLR_HAS_XWAYLAND
+#define class class_t
+#define static
+#include <wlr/xwayland.h>
+#undef static
+#undef class
+#endif
+}
+
+struct process
+{
+    nonstd::observer_ptr<wf::view_interface_t> view;
+    struct wl_client *client;
+    pid_t pid;
+};
+
+static std::map<wf::output_t *, struct process> procs;
+
+class wayfire_background_view : public wf::plugin_interface_t
+{
+    const std::string transformer_name = "background-view";
+    wf::config::option_base_t::updated_callback_t option_changed;
+    wf::option_wrapper_t<std::string> command{"background-view/command"};
+    wf::option_wrapper_t<std::string> file{"background-view/file"};
+    struct wl_event_source *signal = nullptr;
+
+    public:
+    void init() override
+    {
+        grab_interface->name = transformer_name;
+        grab_interface->capabilities = 0;
+
+        option_changed = [=] ()
+        {
+            if (procs[output].view)
+            {
+                procs[output].view->close();
+                kill(procs[output].pid, SIGINT);
+                procs[output].view = nullptr;
+            }
+
+            procs[output].client = client_launch((std::string(command) + add_arg_if_not_empty(file)).c_str());
+        };
+
+        command.set_callback(option_changed);
+        file.set_callback(option_changed);
+
+        output->connect_signal("map-view", &view_mapped);
+
+        if (!signal)
+        {
+            signal = wl_event_loop_add_signal(wf::get_core().ev_loop, SIGCHLD, sigchld_handler, &procs);
+        }
+
+        procs[output].client = client_launch((std::string(command) + add_arg_if_not_empty(file)).c_str());
```

Comment:
Can we not call the updated handler instead of repeating the code?

Reply:
True.